### PR TITLE
Feature/nikhil backend

### PR DESCRIPTION
**Pull request by nikhilpalve18**

### CODE REVIEW

**Review by spreezy-tech — CHANGES_REQUESTED** [review]

**Review by spreezy-tech — COMMENTED** [review]
Excellent job Nikhil on the latest suggestions. You are getting groomed the right way and I can see it in your work. Keep it up :1st_place_medal:

**Review by spreezy-tech — COMMENTED** [review]
Given couple of basic feedback and a practice scenario to help you improve understanding. Excellent work Nikhil

**Review by spreezy-tech — COMMENTED** [review]
Amazing job Nikhil. Im impressed you are able to implement suggestions with ease. Many more things we have here to learn and grow. Keep it up

File: UserRegistrations/src/main/java/com/app/UserRegistration/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
+
+
+    @Operation(summary = "Register the new user", description = "This api helps in registering a new user")
+    @PostMapping("/save-user")
+    public ResponseEntity<?> saveUser(@RequestBody @Valid UserRequestDTO userRequestDTO) throws Exception {
+            this.userService.createUser(userRequestDTO);
+            return ResponseEntity.ok("User created successfully");
```
[the remark this answers is not in the document]
Comment (spreezy-tech) [reply UserRegistrations/src/main/java/com/app/UserRegistration/controller/UserController.java]: That wont be neccessay. But an appropriate response with a generic response DTO will suffice the job

**Review by spreezy-tech — COMMENTED** [review]
Connect with Rohit for multi-threading suggestions. Collaborate and implement it with proper logging so it will be easier to understand for debugging.

File: UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    /* constructor */
+    public UserServiceImpl(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+
+    /* Creating a new user */
+    @Override
+    @Async
+    public CompletableFuture<Void> createUser(UserRequestDTO userRequestDTO) {
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java]: Is there an explicit need for CompletableFuture everywhere?

File: UserRegistrations/src/main/java/com/app/UserRegistration/config/AsyncConfig.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+@Configuration
+@EnableAsync
+public class AsyncConfig {
+
+    @Bean(name = "taskExecutor")
+    public TaskExecutor getTaskExecutor(){
+        ThreadPoolTaskExecutor threadPoolTaskExecutor = new ThreadPoolTaskExecutor();
+        threadPoolTaskExecutor.setCorePoolSize(4);
+        threadPoolTaskExecutor.setMaxPoolSize(64);
+        threadPoolTaskExecutor.setThreadNamePrefix("MyThread: ");
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/config/AsyncConfig.java]: Good job here. Append generated thread Id as well

File: UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java (hunk 2 above)
Comment (nikhilpalve18) [reply UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java]: First I tried without using CompletableFuture then it was unable to handle exceptions properly. Completeable future has a method exceptionally which helps in handling exceptions.

File: UserRegistrations/src/main/java/com/app/UserRegistration/advice/UserExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    @ResponseBody
+    public ResponseEntity<String> handleValidationException(ValidationException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.BAD_REQUEST);
+    }
+    /*
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public UserRegistrationApiResponse handleInvalidInputs(MethodArgumentNotValidException methodArgumentNotValidException){
+        Map<String, String> errors = new HashMap<>();
+        methodArgumentNotValidException.getBindingResult().getFieldErrors().forEach(error->{
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/advice/UserExceptionHandler.java]: Remove commented code if not used

File: UserRegistrations/src/main/java/com/app/UserRegistration/controller/UserController.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return this.userService.getUser(username);
+    }
+
+    @Operation(summary = "Delete user", description = "This api helps in deleting the user by userId")
+    @DeleteMapping("/delete-user/{userId}")
+    public CompletableFuture<ResponseEntity<String>> deleteUser(@PathVariable("userId") Long userId) throws Exception {
+        return this.userService.deleteUser(userId)
+                .thenApply(result -> ResponseEntity.ok("User deleted successfully"))
+                .exceptionally(ex -> {
+                    Throwable cause = (ex.getCause() != null) ? ex.getCause() : ex;
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/controller/UserController.java]: If the exception is occurring in the service, does it go to controller advice or it responds from this exception section?

File: UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+
+    /* Creating a new user */
+    @Override
+    @Async
+    public CompletableFuture<Void> createUser(UserRequestDTO userRequestDTO) throws ValidationException {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        System.out.println("Hello Validator-------------------");
+        userValidator.validate(userRequestDTO);
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java]: Very good

File: UserRegistrations/src/main/java/com/app/UserRegistration/validation/UserValidator.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+    public UserValidator() {
+        this.validator = Validation.buildDefaultValidatorFactory().getValidator();
+    }
+
+    public <T> void validate(T obj) throws ValidationException {
+        Set<ConstraintViolation<T>> violations = validator.validate(obj);
+        if (!violations.isEmpty()) {
+            String errorMessages = violations.stream()
+                    .map(ConstraintViolation::getMessage)
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/validation/UserValidator.java:27]: Good use of map function

Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/validation/UserValidator.java:27]: Good job for using Java 8 way of calling function

**Review by spreezy-tech — CHANGES_REQUESTED** [review]
Excellent job Nikhil. Few comments left. Complete them and we can merge it

File: UserRegistrations/src/main/java/com/app/UserRegistration/controller/UserController.java (hunk 5 above)
Comment (nikhilpalve18) [reply UserRegistrations/src/main/java/com/app/UserRegistration/controller/UserController.java]: The exception wasnt thrown to controllerAdvice it was being handled in controller itself. I reviewed the code and found that exceptionally wasnt necessary if we already have controller advice. You caught it right :)

File: UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        return future;
+    }
+
+
+
+    /* get the details of user by its username */
+    @Override
+    @Async
+    public CompletableFuture<UserResponseDTO> getUser(String username) throws Exception {
+        User user = this.userRepository.findByUsername(username);
```
Comment (spreezy-tech) [inline UserRegistrations/src/main/java/com/app/UserRegistration/service/impl/UserServiceImpl.java]: Use Optional

**Review by spreezy-tech — CHANGES_REQUESTED** [review]
Added comment

**Pull request closed without merging**